### DISCUSSION TITLE
Why isn't `main` the default branch?

### DISCUSSION THREAD

**Discussion by MunifTanjim** [nvim-neo-tree/neo-tree.nvim#414] [Q&A]
Wouldn't it make sense to have the development branch (i.e. `main`) as default?

Creating a new PR defaults to the `v2.x` branch. Contributors need to manually change that.

The general practice is to have the development branch as default. And if somebody wants to pin to a specific version they can do so in their config file.

**Comment by cseickel** [comment]
The reason is that the default branch determines what README is shown, and what is cloned when someone does not specify a branch. In both cases I want that to be the current release and not the development branch that may have bugs or reference unreleased features. There are 1000x more users than contributors, so it makes sense to make their experience the base case.

Note: _The number 1000x is a completely made up guess, but you get the point._

**Reply by MunifTanjim** [reply]
Hmm... I see. I just didn't see this approach anywhere.

> default branch determines what README

Usually I see a notice saying "this is the docs for latest development branch" with a link to the release branch's readme file.

> what is cloned when someone does not specify a branch

From contributors' perspective, they would want the development branch anyway, as they need to create a new branch from `main`.

From the users' perspective, consider this:
- user installs the plugin without specifying any branch (the plugin manager installs `v2.x`)
- maintainer releases `v3.x` and switched the default branch to that
- user updates the plugins and plugin manager fetchs the default branch (e.g. `vim-plug` resets to default branch when running `:PlugUpdate`, doesn't matter which branch was checkout out last)
- now user has broken config, because the installation jumped from `v2.x` to `v3.x`

So even if the intention was to not break user's setup, it can happen anyways. If somebody doesn't want breaking chanages, they should explicity pin to a branch... as implicit pinning won't really work in every scenario (like the example above, or say when the user sets up a different machine with the same config and the plugin manager ends up installing the new `v3.x` branch).

---

But I understand why you wanted to do it in the current way. I just wanted to know the reasons behind it. 👍🏼

**Reply by cseickel** [reply]
That's a good point about users that don't specify a branch. At the very least, they don't get updated until  I determine that it's ready for GA, complete with deprecation warnings meant to guide users to the new configs.

The truth is that I expect contributors to put more effort into it and have more baseline knowledge than a casual drive by "lets try vim" user. The most critical thing to me is that I don't get issues from people using `main` that really shouldn't be using `main`.

> But I understand why you wanted to do it in the current way. I just wanted to know the reasons behind it. 👍🏼

I realize it's a little against convention, but I've never been one to just trust conventional wisdom. 🤷🏼‍♂️

**Reply by MunifTanjim** [reply]
(btw, just to make it clear, I'm not arguing 😂 I'm replying again just to have a discussion around this... not because I want you to change anything.)

> At the very least, they don't get updated until I determine that it's ready for GA, complete with deprecation warnings meant to guide users to the new configs.

Wouldn't that still be unexpected? I mean, as an user:

- If I assume that there won't be any breaking changes even if I don't specify a branch, that would be a wrong assumption.
- If I assume that there would be breaking changes if I don't specify a branch, I would want to face them incrementally (because I want to support the plugin by being an early tester)... but facing a whole bunch of breaking changes together doesn't really help anybody.

So to me it feels like the expectation is not made clear.

> The truth is that I expect contributors to put more effort into it and have more baseline knowledge than a casual drive by "lets try vim" user. The most critical thing to me is that I don't get issues from people using main that really shouldn't be using main.

Well, I can't speak for everybody... but for me, I'm switching to `neo-tree` as my main tree plugin (I was testing it for a while using a separate keybinding before) and I really want to contribute to it. But I don't like that it was intentionally made harder for me to contribute. I don't think discouraging people to make contribution is a good thing.

Instead some other criteria should be set, like:

- PR should container small changes
- PR should have sufficient code coverage
- etc.

And if those criteria are not met, the contributing guideline should make it clear that those PRs would not be accepted and be closed.

> The most critical thing to me is that I don't get issues from people using main that really shouldn't be using main.

That can be prevented by doing the following:

- In the README installation guide, mention the `branch = 'v2.x` so that people following that guide installs from the release branch (✅ this is already there)
- Implement a Issue template that asks for the branch they are using (mention a command that outputs the current branch and ask to copy-paste that output). If that information is missing, automatically close the issue using a bot.

That's a win-win for everybody (user, contributor, maintainer).

---

Sorry for making this long comment 😂 I'm just dumping my thoughts here. Feel free to mark the [first reply](https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/414#discussioncomment-3120544) as "Answer" and lock the conversation.

**Reply by cseickel** [reply]
> Sorry for making this long comment joy I'm just dumping my thoughts here. Feel free to mark the https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/414#discussioncomment-3120544 as "Answer" and lock the conversation.

Nonsense! Conversation is good and this one is thought provoking, I will happily debate these points endlessly. I'll be replying back later...

**Reply by cseickel** [reply]
You know @MunifTanjim, maybe you were right! The readme doesn't change that often and I am constantly correcting PRs. I might just switch to `main` as the default when I move to 3.x.

**Reply by airtonix** [reply]
You could also be instructing users to install against a `neo-tree.nvim@latest` tag.

This is a convention that most docker images and github actions use : 

`neo-tree.nvim@latest > neo-tree.nvim@2 > neo-tree.nvim@2.3 neo-tree.nvim@2.3.5`
